ENH: adding SSA method for IRSA

bsipocz opened the pull request:
Addig SSA access to spectral holdings. The `query_ssa` method is a very thin wrapper of pyvo, so while any issues recovered with it is appreciated, we may ultimately need to propagate those upstream rather than adding workaround for the PR.

closes #3061

keflavich commented:
I'm testing this out now.  See https://github.com/astropy/astroquery/issues/3092 for related work.

First, the format is not the same as `sia`, which ideally we'd fix:
```
In [6]: crd = SkyCoord.from_name('Sgr B2')
   ...: sgrb2_q = Irsa.query_ssa(pos=(crd.ra, crd.dec, 0.01,) )

Traceback (most recent call last):
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:383 in pos
    ra, dec = pos
ValueError: too many values to unpack (expected 2)

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  Cell In[6], line 2
    sgrb2_q = Irsa.query_ssa(pos=(crd.ra, crd.dec, 0.01,) )
  File ~/repos/astroquery/astroquery/ipac/irsa/core.py:160 in query_ssa
    return self.ssa.search(pos=pos, diameter=diameter, band=band, time=time,
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:214 in search
    return self.create_query(
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:260 in create_query
    return SSAQuery(
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:351 in __init__
    self.pos = pos
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:385 in pos
    raise ValueError(
ValueError: Pos must be a sequence with exactly two values, expressing ra and dec in icrs degrees
```
ssa only wants a 2-tuple, not a 3-tuple

But it looks like right now, this is not yet functional (it is labeled `draft`, after all)
```
In [7]: crd = SkyCoord.from_name('Sgr B2')
   ...: sgrb2_q = Irsa.query_ssa(pos=(crd.ra, crd.dec) )
Traceback (most recent call last):
  Cell In[7], line 2
    sgrb2_q = Irsa.query_ssa(pos=(crd.ra, crd.dec) )
  File ~/repos/astroquery/astroquery/ipac/irsa/core.py:160 in query_ssa
    return self.ssa.search(pos=pos, diameter=diameter, band=band, time=time,
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:215 in search
    pos=pos, diameter=diameter, band=band, time=time, format=format, **keywords).execute()
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:578 in execute
    return SSAResults(self.execute_votable(), url=self.queryurl, session=self._session)
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/adhoc.py:111 in __init__
    super().__init__(votable, url=url, session=session)
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/query.py:322 in __init__
    raise DALQueryError(self._status[1], self._status[0], url)
DALQueryError: UsageFault: BAD_REQUEST: Unknown parameter: request

In [8]: crd = SkyCoord.from_name('Sgr B2')
   ...: sgrb2_q = Irsa.query_ssa(pos=crd)
Traceback (most recent call last):
  Cell In[8], line 2
    sgrb2_q = Irsa.query_ssa(pos=crd)
  File ~/repos/astroquery/astroquery/ipac/irsa/core.py:160 in query_ssa
    return self.ssa.search(pos=pos, diameter=diameter, band=band, time=time,
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:215 in search
    pos=pos, diameter=diameter, band=band, time=time, format=format, **keywords).execute()
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/ssa.py:578 in execute
    return SSAResults(self.execute_votable(), url=self.queryurl, session=self._session)
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/adhoc.py:111 in __init__
    super().__init__(votable, url=url, session=session)
  File ~/mambaforge/envs/py310forge/lib/python3.10/site-packages/pyvo/dal/query.py:322 in __init__
    raise DALQueryError(self._status[1], self._status[0], url)
DALQueryError: UsageFault: BAD_REQUEST: Unknown parameter: request
```

bsipocz commented:
Yeap, whatever I have here is just a dump of the first draft that I came up with during some meeting or such by mostly just copying over sia, without cleaning up and making it ssa specific. Realistically this won't get done before the next quarter (and will need some server-side stuff, too, e.g. caching some queries)

bsipocz commented:
@keflavich - this is ready for review. There may be corners of uncovered API, and some missing parsing on the pyvo side, I've opened issues for a couple of those, but the basic examples should work here.

I'm also planning to make this more user friendly, e.g. adding the `to_table()` in the methods, but that will have to be systematic and do in e.g. the alma module, too.

keflavich commented on docs/ipac/irsa/irsa.rst:
```diff
@@ -316,6 +316,87 @@ Now plot the cutout.
+   >>> arp220_spectra = Irsa.query_ssa(pos=coord).to_table()
+
+Without specifying the collection, the query returns results from multiple
+collections. For example this target has spectra from Sofia as well as from
```
SOFIA is an acronym, right?
```suggestion
collections. For example this target has spectra from SOFIA as well as from
```

keflavich commented on astroquery/ipac/irsa/core.py:
```diff
@@ -122,6 +128,42 @@ def query_sia(self, *, pos=None, band=None, time=None, pol=None,
+        pos : `~astropy.coordinates.SkyCoord` class or sequence of two floats
+            the position of the center of the circular search region.
+            assuming icrs decimal degrees if unit is not specified.
+        diameter : `~astropy.units.Quantity` class or scalar float
```
why diameter instead of radius?  radius is more common across the rest of our API

bsipocz replied:
inherited from SSA API, but I agree that I need to change this to be radius instead.

keflavich commented on astroquery/ipac/irsa/core.py:
```diff
@@ -122,6 +128,42 @@ def query_sia(self, *, pos=None, band=None, time=None, pol=None,
+            assuming iso 8601 if format is not specified.
+        collection : str
+           Name of the collection that the data belongs to.
+        format : str
+           the image format(s) of interest.  "all" indicates
+           all available formats; "graphic" indicates
+           graphical images (e.g. jpeg, png, gif; not FITS);
+           "metadata" indicates that no images should be
+           returned--only an empty table with complete metadata.
```
is `('all', 'graphic', 'metadata')` a complete set of valid options?  If so, it would be helpful to indicate that.  If not, can we document what is allowed here or in the main docs?

bsipocz replied:
I copied this from the pyvo docs, looking at the IRSA SSA metadata, these are the actual options:

```
        <OPTION value="ALL"/>
        <OPTION value="COMPLIANT"/>
        <OPTION value="NATIVE"/>
        <OPTION value="VOTABLE"/>
        <OPTION value="FITS"/>
        <OPTION value="XML"/>
        <OPTION value="GRAPHIC"/>
        <OPTION value="METADATA"/>
```

However, some of these are known not to be working (e.g. metadata), and now as I test them some of them are not working as expected. (e.g. FITS). 

So this may need an upstream fix, and here, I may just remove this kwarg for now and hard-wire to return all results in this first iteration.

bsipocz replied:
so, now I hardwired `'all'`, and opened a follow-up to address this properly. xref: #3234

keflavich commented on astroquery/ipac/irsa/tests/test_irsa_remote.py:
```diff
@@ -87,3 +87,10 @@ def test_tap(self):
+        coord = SkyCoord.from_name("Eta Carina")
+        result = Irsa.query_ssa(pos=coord)
+        assert len(result) > 260
+        collections = set(unique(result.to_table(), keys='dataid_collection')['dataid_collection'])
```
is `set(unique(` redundant?  I'd imagine so

bsipocz replied:
sure 🤦‍♀️

keflavich reviewed (commented):
Minor suggestions & questions.  Fine to merge, but it might be nice to make some small doc improvements.

bsipocz commented:
I've addressed all the review issues here and opened follow-up for the format issues. So I'll go ahead with the merge so we can try and sort out if there is anything needed for the next release (in ~2 weeks time).
